File: monitoring-backend/app/schemas/setting.py
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class UserNotificationSettingBase(BaseModel):
    enable_popups: bool = Field(default=True, description="Enable snackbar popups")
    notification_level: str = Field(
        default="all",
        pattern="^(all|warning_critical|critical)$",
        description="Filter level for notifications",
    )


class UserNotificationSettingUpdate(UserNotificationSettingBase):
    pass


class UserNotificationSettingResponse(UserNotificationSettingBase):
    user_id: int
    model_config = ConfigDict(from_attributes=True)


class ThresholdRuleBase(BaseModel):
    metric_type: str = Field(description="e.g., latency, bandwidth_in, bandwidth_out")
    condition: str = Field(
        pattern="^(above|below)$",
        description="Trigger if value goes above or below threshold",
    )
    warning_value: float = Field(ge=0)
    critical_value: float = Field(ge=0)


class ThresholdRuleCreate(ThresholdRuleBase):
    device_type: str


class ThresholdRuleResponse(ThresholdRuleCreate):
    id: int
    model_config = ConfigDict(from_attributes=True)


class SystemConfigBase(BaseModel):
    # Polling
    ping_frequency: int = Field(ge=1, description="Interval in seconds")
    ping_probe_count: int = Field(ge=1)
    ping_timeout_ms: int = Field(ge=100)

    # Sensitivity
    offline_fail_required: int = Field(ge=1)
    recovery_success_required: int = Field(ge=1)
    alert_raise_streak: int = Field(ge=1)
    alert_clear_streak: int = Field(ge=1)

    # Retention
    history_interval_seconds: int = Field(ge=60)
    history_retention_days: int = Field(ge=1)
    alert_retention_days: int = Field(ge=1)


class SystemConfigUpdate(SystemConfigBase):
    pass


class SystemConfigResponse(SystemConfigBase):
    id: int
    model_config = ConfigDict(from_attributes=True)


class BulkSettingsUpdateRequest(BaseModel):
    system_config: SystemConfigUpdate
    threshold_rules: Optional[List[ThresholdRuleCreate]] = None

File: monitoring-frontend/lib/models/setting.dart
class SystemConfig {
  final int pingFrequency;
  final int pingProbeCount;
  final int pingTimeoutMs;
  final int offlineFailRequired;
  final int recoverySuccessRequired;
  final int alertRaiseStreak;
  final int alertClearStreak;
  final int historyIntervalSeconds;
  final int historyRetentionDays;
  final int alertRetentionDays;
  final int topDownMinAlertDurationMinutes;

  SystemConfig(
      {required this.pingFrequency,
      required this.pingProbeCount,
      required this.pingTimeoutMs,
      required this.offlineFailRequired,
      required this.recoverySuccessRequired,
      required this.alertRaiseStreak,
      required this.alertClearStreak,
      required this.historyIntervalSeconds,
      required this.historyRetentionDays,
      required this.alertRetentionDays,
      required this.topDownMinAlertDurationMinutes});

  factory SystemConfig.fromJson(Map<String, dynamic> json) {
    return SystemConfig(
        pingFrequency: json['ping_frequency'] ?? 5,
        pingProbeCount: json['ping_probe_count'] ?? 3,
        pingTimeoutMs: json['ping_timeout_ms'] ?? 1000,
        offlineFailRequired: json['offline_fail_required'] ?? 3,
        recoverySuccessRequired: json['recovery_success_required'] ?? 2,
        alertRaiseStreak: json['alert_raise_streak'] ?? 2,
        alertClearStreak: json['alert_clear_streak'] ?? 2,
        historyIntervalSeconds: json['history_interval_seconds'] ?? 300,
        historyRetentionDays: json['history_retention_days'] ?? 365,
        alertRetentionDays: json['alert_retention_days'] ?? 300,
        topDownMinAlertDurationMinutes:
            json['top_down_min_alert_duration_minutes'] ?? 30);
  }

  Map<String, dynamic> toJson() {
    return {
      'ping_frequency': pingFrequency,
      'ping_probe_count': pingProbeCount,
      'ping_timeout_ms': pingTimeoutMs,
      'offline_fail_required': offlineFailRequired,
      'recovery_success_required': recoverySuccessRequired,
      'alert_raise_streak': alertRaiseStreak,
      'alert_clear_streak': alertClearStreak,
      'history_interval_seconds': historyIntervalSeconds,
      'history_retention_days': historyRetentionDays,
      'alert_retention_days': alertRetentionDays,
      'top_down_min_alert_duration_minutes': topDownMinAlertDurationMinutes
    };
  }
}

class ThresholdRule {
  final String metricType; // 'latency', 'bandwidth_in', 'bandwidth_out'
  final String condition; // 'above', 'below'
  final double warningValue;
  final double criticalValue;

  ThresholdRule({
    required this.metricType,
    required this.condition,
    required this.warningValue,
    required this.criticalValue,
  });

  factory ThresholdRule.fromJson(Map<String, dynamic> json) {
    return ThresholdRule(
      metricType: json['metric_type'] ?? '',
      condition: json['condition'] ?? 'above',
      warningValue: (json['warning_value'] as num?)?.toDouble() ?? 0.0,
      criticalValue: (json['critical_value'] as num?)?.toDouble() ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'metric_type': metricType,
      'condition': condition,
      'warning_value': warningValue,
      'critical_value': criticalValue,
    };
  }
}
